### PR TITLE
MDEV-20392: Detect if 'diff' is not found and needs to be installed

### PR DESCRIPTION
**Pull request by ottok**
Not all environments have 'diff' installed. Most notably CentOS 8
does not have diff out-of-the-box. Thus users running 'cmake .' and
'make' would fail to build MariaDB, and they would think the error
was in ABI incompatibilities due to the error message emitted by CMake
when in reality simply 'diff' was missing.

This fixes is and makes the developer experience better.

## Proof

Running in a clean CentOS 8 container.

### Before

```
$ cmake .
...
-- Looking for cpuid.h
-- Looking for cpuid.h - found
-- Looking for x86intrin.h
-- Looking for x86intrin.h - found
-- Looking for event.h
-- Looking for event.h - not found
-- Configuring done
-- Generating done
-- Build files have been written to: /build

$ make
Scanning dependencies of target abi_check
CMake Error at cmake/do_abi_check.cmake:82 (MESSAGE):
  ABI check found difference between /build/include/mysql/plugin_audit.h.pp
  and /build/abi_check.out

make[2]: *** [CMakeFiles/abi_check.dir/build.make:57: CMakeFiles/abi_check] Error 1
make[1]: *** [CMakeFiles/Makefile2:168: CMakeFiles/abi_check.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

$ yum install -y diffutils
...
Installed:
  diffutils-3.6-6.el8.x86_64
Complete!

$ make
[  0%] Built target abi_check
Scanning dependencies of target INFO_BIN
[  0%] Built target INFO_BIN
Scanning dependencies of target INFO_SRC
[  0%] Built target INFO_SRC
Scanning dependencies of target wsrep_api_v26
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_dummy.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_gtid.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_loader.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_uuid.c.o
[  0%] Linking C static library libwsrep_api_v26.a
[  0%] Built target wsrep_api_v26
```

### After

```
$ make
No such file or directory
CMake Error at cmake/do_abi_check.cmake:80 (MESSAGE):
  Command 'diff' not found.  Please install 'diffutils'.

make[2]: *** [CMakeFiles/abi_check.dir/build.make:57: CMakeFiles/abi_check] Error 1
make[1]: *** [CMakeFiles/Makefile2:168: CMakeFiles/abi_check.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

$ yum install -y diffutils
...
Installed:
  diffutils-3.6-6.el8.x86_64
Complete!

$ make
[  0%] Built target abi_check
Scanning dependencies of target INFO_BIN
[  0%] Built target INFO_BIN
Scanning dependencies of target INFO_SRC
[  0%] Built target INFO_SRC
Scanning dependencies of target wsrep_api_v26
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_dummy.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_gtid.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_loader.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_uuid.c.o
[  0%] Linking C static library libwsrep_api_v26.a
[  0%] Built target wsrep_api_v26
```

### CODE REVIEW

**Review by ericherman — COMMENTED** [review]
I have the sense that it might make sense to check for 'diff' earlier, I don't know cmake but I'm sure cmake has similar facilities of checking for the existance of things as autotools.

Regardless, this is far better than the previous message and I think it could go in as-is. Later it can be improved via prechecking.

**Comment by vuvova** [comment]
@ottok, no, please. This is the rather user-unfriendly way of fixing it.

Instead, skip the whole abi_check if `diff` or `sed` aren't available (let's say `mv` always is).

**Comment by ericherman** [comment]
> Instead, skip the whole abi_check if diff or sed aren't available (let's say mv always is).

@vuvova Interesting! If it's more user-friendly to simply skip the check, why have it at all?

Several questions leap to mind:
* Who needs the abi_check?
* What goes wrong if it fails?
* What are the situations in which it is okay to skip the abi_check?
* Should we typically skip the abi_check?
* When should we never skip the abi_check?

My guess is that your instinct is probably right, because most of the time the abi_check would pass if it could run, but if we skip an abi_check which would have failed if it had run, then a later part of the process will fail based upon the binary incompatibility anyway ....

As you suggest @ottok should change to skipping this check, please help him document why it's okay to do so.

**Comment by vuvova** [comment]
The abi_check rule is supposed to prevent the cases when a developer inadvertently changes API data structures, so that the application loading the new client library would crash. Or an old plugin crashes the new server.

The point is — if one modifies the source code, abi_check rule verifies that outward-looking functions and data structures are not changed.

If one doesn't modify the source files — there is no reason to run the abi_check rule.

**Comment by ericherman** [comment]
So, documentation for the skip could read something like:

```
# By skipping if it is not present, this invites the risk that a developer might somehow both change the code in an ABI-incompatible way AND not have diff installed, thus miss the check ... but this is pretty unlikley.

# Additionally, buildbots will detect and not merget to protected branches
```

**Comment by ottok** [comment]
> The abi_check rule is supposed to prevent the cases when a developer inadvertently changes API data structures, so that the application loading the new client library would crash. Or an old plugin crashes the new server.
> 
> The point is — if one modifies the source code, abi_check rule verifies that outward-looking functions and data structures are not changed.
> 
> If one doesn't modify the source files — there is no reason to run the abi_check rule.

..so you want me to modify the code so that if 'diff' is not available, the whole check is just skipped?

I am fine with any solution as long as we get rid of the misleading error message.


Note that diff is also required by one of the tests. Currently diffutils is a dependency to run the build and run the test suite. Having a proper error message about it would probably be a good service to users.

```
main.mysql_upgrade_noengine 'innodb'     [ fail ]
        Test ended at 2021-06-01 06:16:54
CURRENT_TEST: main.mysql_upgrade_noengine
sh: diff: command not found
The two files differ but it was not possible to execute 'diff' in
order to show only the difference. Instead the whole content of the
two files was shown for you to diff manually.
To get a better report you should install 'diff' on your system, which you
for example can get from http://www.gnu.org/software/diffutils/diffutils.html
 --- /usr/share/mysql-test/main/mysql_upgrade_noengine.result >>>
```

**Comment by ottok** [comment]
Thanks for your review feedback. I will address all actionable feedback and update the PR over the weekend so you can decide to merge this on Monday if you want.

So far the conclusion was that I should rewrite this to "skip diff if not installed, never complain if it is missing".

**Comment by ottok** [comment]
Ok, updated code with a [version](https://github.com/MariaDB/server/compare/b57ea9abadbb0f92a7d40330cd1d7c4ea560cde2..4b9e1e0d926459fac8a0cfe97e0a5d5cdd252a36) that simply skips the check if `diff` is not found. Seems the check is attempted 6 times so the warning is now also emitted 6 times..
```
# make
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
Command 'diff' not found. ABI check skipped.
[  0%] Built target abi_check
Scanning dependencies of target INFO_SRC
[  0%] Built target INFO_SRC
Scanning dependencies of target INFO_BIN
[  0%] Built target INFO_BIN
Scanning dependencies of target wsrep_api_v26
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_dummy.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_gtid.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_loader.c.o
[  0%] Building C object wsrep-lib/wsrep-API/CMakeFiles/wsrep_api_v26.dir/v26/wsrep_uuid.c.o
```

**Comment by vuvova** [comment]
Looks good to merge. In particular, very good that you don't check for `diff` in the main CMakeLists.txt, so `diff` doesn't get into rpm's build-depends.

Three thoughts:
1.  if you don't like six identical messages, you can add the tested abi file name to the message, like
```
Command 'diff' not found. ABI check for mysql.h skipped
```
2. may be, add the same protection for missing `sed` while you're at it? Or is `sed` guaranteed to be present, because it's pulled in by some existing build dependency?
3. please test that is works with non-English locale, like `LANG=fi_FI` for example.

**Comment by ottok** [comment]
I will not implement anything for 'sed'. If sed is missing from the system, you will get a sensible error message already with current code:
```
$ make
CMake Error at cmake/do_abi_check.cmake:73 (MESSAGE):
  sed returned error No such file or directory
```

Sed also seems to be installed out-of-the-box on all distros and sed is not the problem having 1600 hits on Google.

I implemented now a version that prints out:
```
$ make
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_audit.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_ftparser.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/psi/psi_abi_v1.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/psi/psi_abi_v2.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/client_plugin.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_auth.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_password_validation.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_encryption.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_data_type.h skipped.
Command 'diff' not found. ABI check for /build/server/include/mysql/plugin_function.h skipped.
[  0%] Built target abi_check
[  0%] Built target INFO_SRC
[  0%] Built target INFO_BIN
[  0%] Built target wsrep_api_v26
[  0%] Building CXX object wsrep-lib/src/CMakeFiles/wsrep-lib.dir/transaction.cpp.o
^Cmake[2]: *** [wsrep-lib/src/CMakeFiles/wsrep-lib.dir/build.make:264: wsrep-lib/src/CMakeFiles/wsrep-lib.dir/transaction.cpp.o] Unterbrechung
make[1]: *** [CMakeFiles/Makefile2:3156: wsrep-lib/src/CMakeFiles/wsrep-lib.dir/all] Unterbrechung
make: *** [Makefile:182: all] Unterbrechung
```

Tested also with a different LANG set, and works OK.

File: cmake/do_abi_check.cmake
Comment (vuvova) [inline cmake/do_abi_check.cmake:78]: It's quite fragile to use unquoted `${var}` in cmake commands, it'll be simply substituted and split, and a semicolon will break the script. Use either
```cmake
   IF(result MATCHES "No such file or directory")
```
or
```cmake
   IF("${result}" MATCHES "No such file or directory")
```

**Review by vuvova — CHANGES_REQUESTED** [review]

File: cmake/do_abi_check.cmake
Comment (vuvova) [reply cmake/do_abi_check.cmake:78]: _sorry, only noticed it just now_

Comment (ottok) [reply cmake/do_abi_check.cmake]: Fixed in https://github.com/MariaDB/server/compare/0db68e8fe046ad9716ee49d8547023b6a7f16d94..51f5d5fbcae188880be37fea8444d215b38eaadc

Comment (vuvova) [reply cmake/do_abi_check.cmake]: sorry, but you missed the second instance of it, in `ELSEIF(NOT ${result} EQUAL 0)`

Comment (ottok) [reply cmake/do_abi_check.cmake]: True, fixed now both rows, also the one that was there (wrongly) to from before.

**Comment by ottok** [comment]
@vuvova If this is fine, can you please just press "Rebase and merge" so I get the "Github credits" as a thanks for completing this? Thanks

**Comment by grooverdan** [comment]
```
$ sudo mv /usr/bin/diff /usr/bin/nodiff

$ LANG=fi_FI cmake --build . --target abi_check_all --verbose
[1/1] cd /home/dan/repos/build-mariadb-server-10.6 && /usr/bin/cmake -DCOMPILER=/usr/lib64/ccache/clang -DSOURCE_DIR=/home/dan/repos/mariadb-server-10.6 -DBINARY_DIR=/home/dan/repos/build-mariadb-server-10.6 "-DABI_HEADERS=/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_audit.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_ftparser.h;/home/dan/repos/mariadb-server-10.6/include/mysql.h;/home/dan/repos/mariadb-server-10.6/include/mysql/psi/psi_abi_v1.h;/home/dan/repos/mariadb-server-10.6/include/mysql/psi/psi_abi_v2.h;/home/dan/repos/mariadb-server-10.6/include/mysql/client_plugin.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_auth.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_password_validation.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_encryption.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_data_type.h;/home/dan/repos/mariadb-server-10.6/include/mysql/plugin_function.h" -P /home/dan/repos/mariadb-server-10.6/cmake/do_abi_check.cmake
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_audit.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_ftparser.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/psi/psi_abi_v1.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/psi/psi_abi_v2.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/client_plugin.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_auth.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_password_validation.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_encryption.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_data_type.h skipped.
Command 'diff' not found. ABI check for /home/dan/repos/mariadb-server-10.6/include/mysql/plugin_function.h skipped.
```
Works for me.

**Comment by grooverdan** [comment]
> @vuvova If this is fine, can you please just press "Rebase and merge" so I get the "Github credits" as a thanks for completing this? Thanks

152c83d49ca821c54aa49f6b43e33cba63e4d19f was already merged. Please target the earliest active stable branch for bug fixes.

**Comment by vuvova** [comment]
@ottok, yes, sorry. I pushed into 10.2 and your PR was against 10.6. Having pestered you with many small changes for this patch I didn't want to also ask to rebase the PR, especially as I can cherry-pick myself in a minute. If you'd prefer me to rather keep asking you until the commit is ready for "Rebase and merge" I'll do it the future, sure. In particular, note that I also truncated the commit comment. A commit comment should explain the change, but it doesn't have to document the human history since the beginning of the world — having to pagedown multiple times through one commit in `git log` is very annoying.

@grooverdan, please, try to avoid pushing exactly the same change in many main branches, it complicates understanding the history. Although in this case you, probably, did it for Otto to get the "Merged" PR status.

**Comment by ottok** [comment]
@vuvova Yes, please ask contributors to rebase on the correct branch. Giving Github credits is the only award (at least to me) and in general giving contributors a sense of ownership sparks commitment for follow-up work.

It is likely that people (even "seniors") need guidance branch target as contributors might feel it is safer to do changes on in-development branches and not pick everything as a bugfix for a stable release, as some bugfixes causes new bugs and might not always be worth the risk on stable releases. Perhaps there could even be a label on PRs that are targetting "for earliest active stable branch" to make it explicit.

Comment on too verbose documentation noted.

**Pull request merged**
